### PR TITLE
🧠 Feat/logic ask

### PR DESCRIPTION
**Pull request by ccamel**
#### 📝 Purpose 

Here is the first implementation of the logic module that allows for submitting [Prolog](https://en.wikipedia.org/wiki/Prolog) requests through the gRPC service. For the demonstration of the interoperability between the interpreter with its host, the implementation comes with 3 predicates for asking: the name of the chain (`chain_id/1`), the height of the current block (`block_height/1`), and the timestamp of the current block (`block_time`).

#### ⚙️ Technical details

- The implementation is based on the open-source [itchiban/prolog](https://github.com/ichiban/prolog) library - an embeddable prolog interpreter for Go.
- The management of the gas and limits during the execution of the interpreter is achieved through the calculation of a cost for each predicate call. While this approach is not perfect, it does provide a means for fine-tuning the consumption of resources.
- The implementation of the go predicate makes use of context wrapping, which is a bit weird. Another PR will provide some improvements as recomanded in https://github.com/ichiban/prolog/issues/280.

#### CLI

CLI (brought by @bdeneux) gives the following options:

```
okp4d query logic ask --help
Executes the [query] and return the solution(s) found.

Optionally, a program can be transmitted, which will be interpreted before the query is processed.

Since the query is without any side-effect, the query is not executed in the context of a transaction and no fee
is charged for this, but the execution is constrained by the current limits configured in the module (that you can
query).

Usage:
  okp4d query logic ask [query] [flags]

Examples:
$ okp4d query logic ask "chain_id(X)." # returns the chain-id

Flags:
      --grpc-addr string      the gRPC endpoint to use for this chain
      --grpc-insecure         allow gRPC over insecure channels, if not TLS the server must use TLS
      --height int            Use a specific height to query state at (this can error if the node is pruning state)
  -h, --help                  help for ask
      --node string           <host>:<port> to Tendermint RPC interface for this chain (default "tcp://localhost:26657")
  -o, --output string         Output format (text|json) (default "text")
      --program string        reads the program from the given filename or from stdin if "-" is passed as the filename.
      --program-file string   reads the program from the given string.

Global Flags:
      --chain-id string     The network chain ID (default "okp4d")
      --home string         directory for config and data (default "/Users/chris/.okp4d")
      --log_format string   The logging format (json|plain) (default "plain")
      --log_level string    The logging level (trace|debug|info|warn|error|fatal|panic) (default "info")
      --trace               print out full stack trace on errors
```

#### 👨‍💻 Examples

Start the chain, as usual:

```shell
make chain-start
```

Send the following queries:

```shell
$ okp4d query logic ask "chain_id(X)." --output json
{"height":"5199","gas_used":"9086","answer":{"success":true,"has_more":false,"variables":["X"],"results":[{"substitutions":[{"variable":"X","term":{"name":"okp4-localnet","arguments":[]}}]}]}}

$ okp4d query logic ask "block_height(H), block_time(T)." --output json
{"height":"5214","gas_used":"9087","answer":{"success":true,"has_more":false,"variables":["H","T"],"results":[{"substitutions":[{"variable":"H","term":{"name":"5214","arguments":[]}},{"variable":"T","term":{"name":"1673354337","arguments":[]}}]}]}}
```

### CODE REVIEW

**Review by amimart — CHANGES_REQUESTED** [review]

**Review by bdeneux — APPROVED** [review]
Seems good to me and it's working nice on my side  👍

File: x/logic/keeper/grpc_query_ask_test.go.old [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package keeper
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/keeper/grpc_query_ask_test.go.old:1]: ah ah, good catch! 😁

**Review by amimart — APPROVED** [review]
Mind blowing!

**Comment by ccamel** [comment]
Here we go! Thanks :heart:

**Pull request merged**